### PR TITLE
Addition of LLVM.

### PR DESCRIPTION
**Pull request by dada2048**
This is a proposed addition of the LLVM Compiler Infrastructure (version 4.0.0).
It compiles and works for at least one example program but it needs more testing and maybe some changes (this is my first port, I'm not sure I did everything correctly).

Port request is in issue #508.

### CODE REVIEW

File: ports/llvm/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/lib/cmake)
+
+# Remove extra copies of include files in debug directory
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+
+# Remove bin directories
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/bin)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/bin)
+
+# Remove one empty include subdirectory
```
Comment (dada2048) [inline ports/llvm/portfile.cmake]: Is there a safer way to do this? Any upstream change that populates this directory will require a change to the port...

File: ports/llvm/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    OPTIONS
+        -DLLVM_TARGETS_TO_BUILD=X86
+        -DLLVM_BUILD_TOOLS=OFF
+        -DLLVM_BUILD_UTILS=OFF
+        -DLLVM_ABI_BREAKING_CHECKS=FORCE_OFF
+)
+
+vcpkg_install_cmake()
+
+# Move cmake modules to correct vcpkg location and remove extra copy
```
Comment (dada2048) [inline ports/llvm/portfile.cmake]: I'm not sure those cmake files can handle being installed in a different location, needs testing.

File: ports/llvm/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+file(COPY ${CURRENT_PACKAGES_DIR}/lib/cmake/llvm DESTINATION ${CURRENT_PACKAGES_DIR}/share)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/lib/cmake)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/lib/cmake)
+
+# Remove extra copies of include files in debug directory
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+
```
Comment (dada2048) [inline ports/llvm/portfile.cmake]: Is there a safer way to do this? Any upstream change that populates this directory will require a change to the port...

File: ports/llvm/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        -DLLVM_BUILD_UTILS=OFF
+        -DLLVM_ABI_BREAKING_CHECKS=FORCE_OFF
+)
+
+vcpkg_install_cmake()
+
+# Move cmake modules to correct vcpkg location and remove extra copy
+file(COPY ${CURRENT_PACKAGES_DIR}/lib/cmake/llvm DESTINATION ${CURRENT_PACKAGES_DIR}/share)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/lib/cmake)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/lib/cmake)
```
Comment (dada2048) [inline ports/llvm/portfile.cmake]: I'm not sure those cmake files can handle being installed in a different location, needs testing.

File: ports/llvm/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+# Remove extra copies of include files in debug directory
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+
+# Remove bin directories
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/bin)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/bin)
+
+# Remove one empty include subdirectory
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/include/llvm/MC/MCAnalysis)
```
Comment (dada2048) [inline ports/llvm/portfile.cmake]: Is there a safer way to do this? Any upstream change that populates this directory will require a change to the port...

File: ports/llvm/portfile.cmake (hunk 4 above)
Comment (ras0219-msft) [inline ports/llvm/portfile.cmake]: We do have a helper script that performs a bunch of automatic munging as long as the files are pretty standard -- `vcpkg_fixup_cmake_targets()`.

In this case, I think these three lines can be replaced with
```cmake
vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/llvm)
```

File: ports/llvm/portfile.cmake (hunk 5 above)
Comment (ras0219-msft) [inline ports/llvm/portfile.cmake]: We could check if there are contents in the folder before deleting it:
```cmake
file(GLOB X ${CURRENT_PACKAGES_DIR}/include/llvm/MC/MCAnalysis/*)
if(NOT X)
  file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/include/llvm/MC/MCAnalysis)
endif()
```

File: ports/llvm/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    FILENAME "llvm-4.0.0.src.tar.xz"
+    SHA512 cf681f0626ef6d568d951cdc3e143471a1d7715a0ba11e52aa273cf5d8d421e1357ef2645cc85879eaefcd577e99e74d07b01566825b3d0461171ef2cbfc7704
+)
+vcpkg_extract_source_archive(${ARCHIVE})
+
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS
+        -DLLVM_TARGETS_TO_BUILD=X86
```
Comment (ras0219-msft) [inline ports/llvm/portfile.cmake]: Does this include the entire x86 family (+amd64)? Or is this just i686?

We might want to think about ARM here as well, but that's a longer term issue -- it shouldn't block this initial port.

Comment (dada2048) [inline ports/llvm/portfile.cmake]: This is just 32bit X86. I'm open to make it "X86;AArch64" or "X86;AArch64;ARM" although I'm having trouble with the syntax for passing quotes.

But I was actually planning to leave out this option altogether, which builds all architectures (that's the default on all platforms). The reason I have X86 now is that it takes a horrendous time to compile and I wanted to iterate a bit faster.

I will implement your other suggestions this weekend and do some more testing (already have a couple examples running with JIT and everything so I believe I'm on the right track).

**Comment by dada2048** [comment]
I think this is now ready to merge.

Feel free to remove the LLVM_TARGETS_TO_BUILD option if you prefer to have all architectures built by default but the buildtree is already twice the size of boost with just x86 and x64...

**Comment by dada2048** [comment]
Anything else I can do here?

**Comment by ras0219-msft** [comment]
Nope, thanks for the PR!

(Sorry for the delay! It's summer vacation season in the US 🌴)

**Comment by dada2048** [comment]
No worries! Thanks for the feedback and for merging.

**Comment by AraHaan** [comment]
Was this updated for clang 5.0.0 too? I need it to work with 5.0.0.

**Pull request merged**